### PR TITLE
Clean up hipDNN batch norm for mergeability

### PR DESCRIPTION
**Pull request by zjgarvey**
## Summary
- Replace MIOpen with hipDNN graph API for all three batch norm paths (training forward, inference, backward) in hipDNN backend.
- Remove debug prints (`std::cout`) and commented-out MIOpen code throughout the hipDNN integration files
- Fix backward dispatch by giving hipDNN its own `impl_index` (3) in `Normalization.cpp`, instead of sharing index 2 with MIOpen
- Use `batchnorm_inference_variance_ext()` for inference, which correctly handles variance+epsilon internally
- Add workspace allocation for all graph API paths
- Revert unrelated cmake workarounds (hardcoded aotriton version, removed rocsolver, moved fmt)
- Detect hipDNN during configure to set `USE_HIPDNN` and determine which backend to use. (need to add an extra flag during backend selection, or set default to false for hipDNN until it is stable).
- Adds a minimal python `torch.backends.hipdnn`.

### CODE REVIEW

File: aten/src/ATen/native/Normalization.cpp [hunk 1]
```diff
@@ -530,7 +530,11 @@ BatchNormBackend _select_batch_norm_backend(
   bool is_miopen_3_4 = miopen_version >= 30400;  // ROCm 6.4
   bool is_miopen_3_5 = miopen_version >= 30500;  // ROCm 7.0
   bool PYTORCH_MIOPEN_SUGGEST_NHWC_BATCHNORM = c10::utils::check_env("PYTORCH_MIOPEN_SUGGEST_NHWC_BATCHNORM").value_or(is_miopen_3_5);
-  bool PYTORCH_ENABLE_HIPDNN_BATCHNORM = c10::utils::check_env("PYTORCH_ENABLE_HIPDNN_BATCHNORM").value_or(false);
+#ifdef USE_HIPDNN
+  bool hipdnn_enabled = at::globalContext().userEnabledHipdnn();
+#else
+  bool hipdnn_enabled = false;
+#endif
```
Comment (zjgarvey) [inline aten/src/ATen/native/Normalization.cpp]: Note: this is currently removing the environment variable and just checking if hipDNN is enabled or not. I think this makes the most sense, because we can just call `torch.miopen_batch_norm` if we need to pick a specific one for testing, no?

File: aten/src/ATen/native/hipdnn/BatchNorm_hipdnn.cpp [hunk 2]
```diff
@@ -112,149 +124,115 @@ std::tuple<Tensor, Tensor, Tensor> hipdnn_batch_norm(
...
-  auto handle = getMiopenHandle();
-  auto dataType = getMiopenDataType(*input);
-  TensorDescriptor idesc{ *input, 4 };  // input descriptor
-  TensorDescriptor wdesc{ expandScale(*weight, input->dim()), 4 };  // descriptor for weight, bias, running_mean, etc.
-
-  Constant one(dataType, 1);
-  Constant zero(dataType, 0);
+  auto handle = getHipdnnHandle();
+  auto inputType = getHipdnnDataType(*input);
+  auto intermediateType = getHipdnnDataType(*weight);
```
Comment (zjgarvey) [inline aten/src/ATen/native/hipdnn/BatchNorm_hipdnn.cpp:132]: Check this.

File: aten/src/ATen/native/hipdnn/BatchNorm_hipdnn.cpp [hunk 3]
```diff
@@ -112,149 +124,115 @@ std::tuple<Tensor, Tensor, Tensor> hipdnn_batch_norm(
...
+    variantPack[weight_attr->get_uid()] = weight->data_ptr();
+    variantPack[bias_attr->get_uid()] = bias->data_ptr();
+    variantPack[y->get_uid()] = output->data_ptr();
+    variantPack[savedMean->get_uid()] = save_mean.data_ptr();
+    variantPack[savedInvVar->get_uid()] = save_var.data_ptr();
+    if (running_mean->defined()) {
+      variantPack[prev_mean_attr->get_uid()] = running_mean->data_ptr();
+      variantPack[prev_var_attr->get_uid()] = running_var->data_ptr();
+      variantPack[nextMean->get_uid()] = running_mean->data_ptr();
+      variantPack[nextVar->get_uid()] = running_var->data_ptr();
```
Comment (zjgarvey) [inline aten/src/ATen/native/hipdnn/BatchNorm_hipdnn.cpp:192]: Shared data for prev and next? Did we make a conclusion on this?

File: cmake/Dependencies.cmake [hunk 4]
```diff
@@ -1597,7 +1599,7 @@ endif()
 set(FMT_INSTALL ON)
 set(TEMP_BUILD_SHARED_LIBS ${BUILD_SHARED_LIBS})
 set(BUILD_SHARED_LIBS OFF CACHE BOOL "Build shared libs" FORCE)
-# add_subdirectory(${PROJECT_SOURCE_DIR}/third_party/fmt)
+add_subdirectory(${PROJECT_SOURCE_DIR}/third_party/fmt)
```
Comment (zjgarvey) [inline cmake/Dependencies.cmake:1602]: Why was this commented out in your branch?

File: aten/src/ATen/native/hipdnn/BatchNorm_hipdnn.cpp [hunk 5]
```diff
@@ -112,149 +124,115 @@ std::tuple<Tensor, Tensor, Tensor> hipdnn_batch_norm(
...
+    if (running_mean->defined()) {
+      nextMean->set_output(true).set_data_type(intermediateType);
+      nextVar->set_output(true).set_data_type(intermediateType);
+    }
+
+    HIPDNN_FE_CHECK(graph->build(handle));
+
+    int64_t workspace_size = 0;
+    HIPDNN_FE_CHECK(graph->get_workspace_size(workspace_size));
+    auto workspace = at::empty({workspace_size}, input_t.options().dtype(at::kByte));
```
Comment (zjgarvey) [inline aten/src/ATen/native/hipdnn/BatchNorm_hipdnn.cpp:179]: Bring this up with Jeff? How should we allocate workspace?

Comment (zjgarvey) [reply aten/src/ATen/native/hipdnn/BatchNorm_hipdnn.cpp:179]: @jeffdaily

File: aten/src/ATen/native/hipdnn/BatchNorm_hipdnn.cpp [hunk 6]
```diff
@@ -112,149 +124,115 @@ std::tuple<Tensor, Tensor, Tensor> hipdnn_batch_norm(
...
-    auto invVariance_attr = createTensorAttributes(expandScale(*running_var, input->dim()));
+    auto variance_attr = createTensorAttributes(expandScale(*running_var, input->dim()));
+    auto epsilon_attr = std::make_shared<hipdnn_frontend::graph::TensorAttributes>();
+    epsilon_attr->set_value(epsilon);
 
-    auto output_attr = graph->batchnorm_inference(
-      input_attr, mean_attr, invVariance_attr, weight_attr, bias_attr, bnAttributes);
+    auto bnAttributes = hipdnn_frontend::graph::BatchnormInferenceAttributesVarianceExt();
+    auto output_attr = graph->batchnorm_inference_variance_ext(
+        input_attr, mean_attr, variance_attr, weight_attr, bias_attr, epsilon_attr, bnAttributes);
```
Comment (zjgarvey) [inline aten/src/ATen/native/hipdnn/BatchNorm_hipdnn.cpp:216]: Check which inference we should use in this case?

File: aten/src/ATen/native/hipdnn/BatchNorm_hipdnn.cpp [hunk 7]
```diff
@@ -304,55 +279,51 @@ std::tuple<Tensor, Tensor, Tensor> hipdnn_batch_norm_backward(
...
+  auto graph = std::make_shared<hipdnn_frontend::graph::Graph>();
+  graph->set_io_data_type(inputType)
+      .set_intermediate_data_type(intermediateType)
+      .set_compute_data_type(hipdnn_frontend::DataType::FLOAT);
+
+  auto dy_attr = createTensorAttributes(*grad_output);
+  auto input_attr = createTensorAttributes(*input);
+  auto weight_attr = createTensorAttributes(expandScale(*weight, input->dim()));
+  auto savedMeanAttr = createTensorAttributes(expandScale(*save_mean, input->dim()));
+  auto savedInvVarAttr = createTensorAttributes(expandScale(*save_var, input->dim()));
```
Comment (zjgarvey) [inline aten/src/ATen/native/hipdnn/BatchNorm_hipdnn.cpp:299]: Verify the expand scale use here.

File: aten/src/ATen/native/hipdnn/BatchNorm_hipdnn.cpp (hunk 6 above)
Comment (zjgarvey) [reply aten/src/ATen/native/hipdnn/BatchNorm_hipdnn.cpp:216]: Note: `batchnorm_inference` expects a pre-computed inverse variance `(1 / sqrt(var + eps))`, meaning the caller must transform `running_var` before passing it. On the other hand, `batchnorm_inference_variance_ext` takes raw variance and epsilon separately, and does the `1 / sqrt(var + eps)` internally. This change is good.

File: aten/src/ATen/native/hipdnn/BatchNorm_hipdnn.cpp (hunk 7 above)
Comment (zjgarvey) [reply aten/src/ATen/native/hipdnn/BatchNorm_hipdnn.cpp:299]: Yep, this is correct.
